lhotse/dataset/collation.py

```diff
@@ -8,19 +8,115 @@
+        unk_symbol: str = "<unk>",
+    ):
+        assert all(len(cut.supervisions) == 1 for cut in cuts), \
+            "Expects only Cuts with a single supervision"
+
+        self.pad_symbol = pad_symbol
+        self.bos_symbol = bos_symbol
+        self.eos_symbol = eos_symbol
+        self.unk_symbol = unk_symbol
+
+        self.add_eos = add_eos
+        self.add_bos = add_bos
+
+        tokens = {
+            char for cut in cuts
+            for char in cut.supervisions[0].text
+        }
+        tokens_unique = (
+            [pad_symbol, unk_symbol]
+            + ([bos_symbol] if add_bos else [])
+            + ([eos_symbol] if add_eos else [])
+            + sorted(tokens)
+        )
+
+        self.token2idx = {token: idx for idx, token in enumerate(tokens_unique)}
+        self.idx2token = [token for token in tokens_unique]
+
+    def __call__(self, cuts: CutSet) -> Tuple[torch.Tensor, torch.Tensor]:
+
+        assert all(len(cut.supervisions) == 1 for cut in cuts), \
+            "Expects only Cuts with a single supervision"
+
+        token_sequences = [cut.supervisions[0].text for cut in cuts]
+        max_len = len(max(token_sequences, key=len))
+
+        seqs = [
+            ([self.bos_symbol] if self.add_bos else [])
+            + list(seq)
+            + ([self.eos_symbol] if self.add_eos else [])
+            + [self.pad_symbol] * (max_len - len(seq))
+            for seq in token_sequences
+        ]
+
+        tokens_batch = torch.from_numpy(np.array([
+            [self.token2idx[token] for token in seq]
+            for seq in seqs
+        ], dtype=np.int))
+
+        tokens_lens = torch.IntTensor([
+            len(seq) + int(self.add_eos) + int(self.add_bos)
+            for seq in token_sequences
+        ])
+
+        return tokens_batch, tokens_lens
+
+    def inverse(self, tokens_batch: torch.IntTensor, tokens_lens: torch.IntTensor) -> List[str]:
+        start = 1 if self.add_bos else 0
+        sentences = [
+            "".join([
+                self.idx2token[idx]
+                for idx in tokens_list[start:end - int(self.add_eos)]
+            ])
+            for tokens_list, end in zip(tokens_batch, tokens_lens)
+        ]
+        return sentences
+
+
+def collate_features(cuts: CutSet) -> Tuple[torch.Tensor, torch.IntTensor]:
     """
     Load features for all the cuts and return them as a batch in a torch tensor.
     The output shape is ``(batch, time, features)``.
     The cuts will be padded with silence if necessary.
     """
     assert all(cut.has_features for cut in cuts)
+    features_lens = torch.tensor([cut.num_frames for cut in cuts], dtype=torch.int)
     cuts = maybe_pad(cuts)
     first_cut = next(iter(cuts))
     features = torch.empty(len(cuts), first_cut.num_frames, first_cut.num_features)
     for idx, cut in enumerate(cuts):
         features[idx] = torch.from_numpy(cut.load_features())
-    return features
+    return features, features_lens
```

Comment:
Do you mind updating the rest of `collate_X` functions? There is an API mismatch at the moment, and they would have also benefitted from your changes.

Reply:
Good point, I am off the computer already, so I will push an update tomorrow :)

Reply:
I think at the very least add lengths to collate audio below, and update TTS dataset to have `audio_lens` too. Multi-channel collations are ok the be left behind for now. 

I'll merge after that. Thanks!

Reply:
I added the audio_lens, adjusted test for the unsupervised audio dataset and added audio_lens to the speech synthesis dataset 👍🏻